smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java

```diff
@@ -0,0 +1,364 @@
+            }
+
+
+            for (String emoji : emojis) {
+                if (!restrictions.getAllowedEmojis().contains(emoji)) {
+                    throw new IllegalArgumentException("Emoji " + emoji + " is not allowed");
+                }
+            }
+        }
+
+        List<Reaction> reactions = new ArrayList<>();
+        for (String emoji : emojis) {
+            reactions.add(new Reaction(emoji));
+        }
+
+        ReactionsElement reactionsElement = new ReactionsElement(reactions, originalMessageId);
+
+        MessageBuilder newMessage = StanzaBuilder.buildMessage();
+
+        newMessage.addExtension(reactionsElement);
+
+        return newMessage;
+    }
+
+    /**
+     * Adds a reactions listener to the collection.
+     * <p>
+     * This method ensures that the listener is only added if it is not already present in the set.
+     * It is synchronized to ensure thread safety when adding listeners.
+     *
+     * @param listener The reactions listener to be added to the set.
+     * @return {@code true} if the listener was successfully added;
+     *         {@code false} if the listener was already present and not added.
+     */
+    public synchronized boolean addReactionsListener(ReactionsListener listener) {
+        return listeners.add(listener);
+    }
+
+    /**
+     * Removes a reactions listener from the collection.
+     * <p>
+     * This method ensures that the listener is removed from the set, if it exists.
+     * It is synchronized to ensure thread safety when removing listeners.
+     *
+     * @param listener The reactions listener to be removed from the set.
+     * @return {@code true} if the listener was successfully removed;
+     *         {@code false} if the listener was not found in the set.
+     */
+    public synchronized boolean removeReactionsListener(ReactionsListener listener) {
+        return listeners.remove(listener);
+    }
+
+    /**
+     * Creates a form for reaction restrictions, including the max number of reactions per user
+     * and the list of allowed emojis.
+     *
+     * @param maxReactionsPerUser The maximum number of reactions allowed per user.
+     * @param allowedEmojis The list of allowed emojis.
+     * @return The reaction restrictions form.
+     */
+    public DataForm createReactionRestrictionsForm(int maxReactionsPerUser, List<String> allowedEmojis) {
+
+        DataForm.Builder builder = DataForm.builder();
+        builder.setFormType(String.valueOf(DataForm.Type.result));
+
+        builder.addField(
+                        FormField.builder("max_reactions_per_user").setValue(String.valueOf(maxReactionsPerUser))
+                                        .build()
+        );
+
+        FormField.Builder<TextSingleFormField, TextSingleFormField.Builder> allowlistFieldBuilder = FormField.builder("allowlist");
+        for (String emoji : allowedEmojis) {
+            Reaction reaction = new Reaction(emoji);
```

Comment:
Why do you wrap the emoji in a `Reaction` just to extract the emoji again one line later?